### PR TITLE
[Bugfix] Propagate VLLM_MARLIN_INPUT_DTYPE into INC/AutoRound quant path

### LINKED ISSUE #48904: [Bug]: VLLM_MARLIN_INPUT_DTYPE=int8 silently ignored for auto-round (INC) checkpoints
### Your current environment

<details>
<summary>vLLM v0.25.1, RTX 3090 (Ampere, sm_86), CUDA 12/13</summary>

Reproduces on any Ampere GPU serving an **auto-round**-format checkpoint.
</details>

### 🐛 Describe the bug

`VLLM_MARLIN_INPUT_DTYPE=int8` (the Marlin W4A8 int8-activation toggle added in #24722) is **silently ignored** for checkpoints in the `auto-round` format. The layer keeps running W4A16 — no int8 activation quantization, no error, and prefill throughput is unchanged.

It works correctly for GPTQ/AWQ checkpoints; only `auto-round` is affected.

**Root cause.** `auto-round` checkpoints are routed to the `inc` quantization path (`INCConfig.override_quantization_method`). That path builds the linear method in `inc/schemes/inc_wna16_linear.py::_build_gptq_method`:

```python
return AutoGPTQLinearMethod(
    AutoGPTQConfig(...)
)
```

but it never assigns `input_dtype`. The equivalent GPTQ/AWQ path in `auto_gptq.py::get_quant_method` does:

```python
quant_method.input_dtype = get_marlin_input_dtype(prefix)   # auto_gptq.py:269 (and :261 for MoE)
```

`AutoGPTQLinearMethod.__init__` defaults `self.input_dtype = None`, so without that assignment the marlin kernel never enters the int8-activation branch and the toggle is a no-op.

### Reproduce

```bash
# any auto-round int4 W4A16 checkpoint (quant_method: auto-round in config.json)
VLLM_MARLIN_INPUT_DTYPE=int8 vllm serve <auto-round-w4a16-model> --quantization auto_round --dtype bfloat16
```

Observed: boot logs `Using MarlinLinearKernel for AutoGPTQLinearMethod`, generation is coherent, but prefill TPS is identical to a run **without** the env var (no int8 acceleration). The same env var on a GPTQ checkpoint of the same base gives the expected int8 prefill speedup.

### Suggested fix

Assign `input_dtype` after constructing `AutoGPTQLinearMethod` in `_build_gptq_method`, mirroring `auto_gptq.py`:

```python
method = AutoGPTQLinearMethod(AutoGPTQConfig(...))
method.input_dtype = get_marlin_input_dtype()
return method
```

(A MoE analog exists in `inc_wna16_scheme.py::get_moe_method`, which builds `AutoGPTQMoEMethod` without threading `input_dtype` either.)

### PR DESCRIPTION
**Pull request by lazypool**
Fixes #48904.

AutoRound (INC) checkpoints route through `inc_wna16_linear.py` and `inc_wna16_scheme.py`, which construct Marlin linear/MoE methods but never assign `input_dtype`. The attribute defaults to `None`, so the Marlin kernel never enters the int8-activation branch regardless of the `VLLM_MARLIN_INPUT_DTYPE` env var.

Fix by setting `method.input_dtype = get_marlin_input_dtype()` after construction, matching the pattern already used in `auto_gptq.py:269` and `auto_gptq.py:261` for the GPTQ/AWQ path.

## Verification
- Syntax validated by `ruff check` and `ruff format --check`
- Logic matches the working GPTQ/AWQ path byte-for-byte
- @kevinb361 independently confirmed the same fix enables W4A8 on 2x RTX 3090 with 1.36–1.61x TTFT speedup in [his comment](#issuecomment-5007553020)

Co-authored-by: noonghunna <10742901+noonghunna@users.noreply.github.com>
Co-authored-by: kevinb361 <91768207+kevinb361@users.noreply.github.com>

### CODE REVIEW

**Comment by kevinb361** [comment]
I validated the exact #49039 commit on a real 2× RTX 3090 AutoRound deployment using vLLM 0.25.1, with the separate negative-scale workaround from #48905 applied.

Startup confirmed:
- `VLLM_MARLIN_INPUT_DTYPE=int8`
- `MarlinLinearKernel` selected on both TP workers
- AutoRound negative group-scale folding engaged

A controlled 550,000-character prefill completed coherently:
- TTFT: 47.566 s
- wall time: 50.833 s
- output: coherent one-sentence summary

This confirms the INC/AutoRound dtype propagation works in the real serving path. This run was not a matched W4A16 A/B, so I’m not treating it as a standalone speedup measurement. The negative-scale issue remains a separate prerequisite for this checkpoint.

**Comment by hebo1221** [comment]
I tested this on a GB10 while following up on #49546. Applying the diff from `7ad8269836ef24872de675b88fcfcfb1a9b569f8` to my PyPI vLLM 0.29.0 install exposed two reversed assertions in `AutoGPTQMoEMethod`. Both currently require 8-bit weights for 8-bit activations, which rejects this INT4 checkpoint. The changes needed in that install are:

```diff
# create_weights
- assert self.quant_config.quant_type.size_bits == 8
+ assert self.quant_config.quant_type.size_bits == 4

# process_weights_after_loading
- assert not is_a_8bit or self.quant_config.quant_type.size_bits == 8
+ assert not is_a_8bit or self.quant_config.quant_type.size_bits == 4
```

With your propagation change and those two corrections, the model loads and the corrupted output disappears on my four test prompts. I checked that the INC linear kernel configuration and both the MoE preparation and execution paths use `torch.float8_e4m3fn` in that run.

The stock FP8-option run had `AutoGPTQMoEMethod.input_dtype=None` but `MarlinExperts.input_dtype=float8_e4m3fn`. As a separate control, I left the prepared weights alone and changed only the runtime dtype on all 48 MoE backends to match the preparation dtype. That restored coherent output too, without a CUDA change.

Environment: GB10 / SM121, driver 580.126.09, PyTorch 2.13.0+cu130; `Intel/Qwen3.5-122B-A10B-int4-AutoRound` at `3045d02bb737effc4581da91bddbad3be02934e4`. TP=1, eager mode, no speculative decoding or prefix caching, context 2048, concurrency 1, KV cache 2 GiB. Generation used temperature=0, seed=0 and max_tokens=256.

This is a reproduction check, not a quality or speed benchmark. Three outputs reached the token limit; the capital question ended with Paris. The patched outputs are not all token-identical to BF16.

I also ran the six CPU checks below: **6 failed with the original guards, 6 passed with the corrections**. They cover INT4 weight creation with INT8/FP8 activations and rejection of unsupported INT8-weight/8-bit-activation combinations. The successful FP8 CUDA inference was a separate full-model run.

<details>
<summary>Regression checks and command</summary>

Save as `test_gptq_moe_a8_contract.py` outside the source checkout, then run in the vLLM environment:

```sh
python -m pytest --confcutdir=. test_gptq_moe_a8_contract.py -q
```

```python
"""Regression checks for the W4A8 vs unsupported W8A8 MoE contract.

These use the same constructor-free setup as upstream's existing
test_auto_gptq_moe_creates_zero_initialized_expert_biases.
The end-to-end GB10 model runs separately cover successful weight processing.
"""
from types import SimpleNamespace

import pytest
import torch

from vllm.model_executor.layers.quantization.auto_gptq import (
    AutoGPTQConfig,
    AutoGPTQMoEMethod,
)


def method_for(bits, activation):
    method = object.__new__(AutoGPTQMoEMethod)
    method.quant_config = AutoGPTQConfig(bits, 128, False, True, False, {}, {})
    method.input_dtype = activation
    method.experts_cls = None
    method.moe = SimpleNamespace(w13_num_shards=2)
    return method


def create(method):
    layer = torch.nn.Module()
    method.create_weights(
        layer, 2, 256, 128, torch.bfloat16,
        intermediate_size_full=128, weight_loader=lambda *args, **kwargs: None,
    )
    return layer


@pytest.mark.parametrize('activation', [torch.int8, torch.float8_e4m3fn])
def test_int4_moe_accepts_eight_bit_activation(activation):
    layer = create(method_for(4, activation))
    assert layer.w13_qweight.shape == (2, 32, 256)
    assert layer.w2_qweight.shape == (2, 16, 256)


@pytest.mark.parametrize('activation', [torch.int8, torch.float8_e4m3fn])
@pytest.mark.parametrize('operation', ['create', 'process'])
def test_int8_moe_rejects_unsupported_eight_bit_activation(activation, operation):
    method = method_for(8, activation)
    with pytest.raises(AssertionError, match='W8A8-INT8 is not supported'):
        if operation == 'create':
            create(method)
        else:
            method.process_weights_after_loading(torch.nn.Module())
```

</details>

**Comment by lazypool** [comment]
@hebo1221 Thanks for the thorough investigation! Good catch on the reversed assertions. I've added a second commit with the fix and listed you as co-author. Feel free to test on your GB10 anytime ~

and @tobby168, I hope this PR helps with your issue #49546 ! 😊
